# Support Vector Machine (SVM)

In [1]:

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [2]:
# Importing prepared dataset
import pickle
input = open('data_LogRegr.pkl', 'rb')
obj = pickle.load(input)
input.close()
X_train = obj["X_train"]
X_test = obj["X_test"]
y_train = obj["y_train"]
y_test = obj["y_test"]

In [3]:

# Fitting SVM to the Training set 
from sklearn.svm import SVC
svm = SVC(kernel = 'rbf', random_state = 10).fit(X_train, y_train)

In [4]:

# Predicting the Test set results
y_pred = svm.predict(X_test)
svm.score(X_test,y_test)

0.8416666666666667

In [5]:
# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)
print(cm)

[[1076  116]
 [ 150  338]]


Распознано неверно 266 обьектов (при радиальном ядре)

In [6]:
# Fitting SVM to the Training set
from sklearn.svm import SVC
svm2 = SVC(kernel = 'sigmoid', random_state = 10).fit(X_train, y_train)

In [7]:
# Predicting the Test set results
y_pred = svm2.predict(X_test)
svm2.score(X_test,y_test)

0.7964285714285714

In [8]:
# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm2 = confusion_matrix(y_test, y_pred)
print(cm2)

[[1019  173]
 [ 169  319]]


Распознано неверно 342 обьектов (при сигмоидальном ядре)

In [9]:
# Fitting SVM to the Training set 
from sklearn.svm import SVC
svm3 = SVC(kernel = 'poly', random_state = 10).fit(X_train, y_train)

In [10]:
# Predicting the Test set results
y_pred = svm3.predict(X_test)
svm3.score(X_test,y_test)

0.8422619047619048

In [11]:
# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm3 = confusion_matrix(y_test, y_pred)
print(cm3)

[[1106   86]
 [ 179  309]]


Распознано неверно 265 обьектов (при полиномиальном ядре)

In [12]:
# Fitting SVM to the Training set
from sklearn.svm import SVC
svm4 = SVC(kernel = 'linear', random_state = 10).fit(X_train, y_train)

In [13]:
# Predicting the Test set results
y_pred = svm4.predict(X_test)
svm4.score(X_test,y_test)

0.8505952380952381

In [14]:
# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix
cm4 = confusion_matrix(y_test, y_pred)
print(cm4)

[[1070  122]
 [ 129  359]]


Распознано неверно 251 обьект (линейное ядро). 

Для данной модели проведены эксперименты с ядрами. 

ВЫВОДЫ: обратим внимание на модель с полиномиальным ядром. Как мы видим, в ней верно распознано 1106 людей, которые проживают за чертой бедности (иневерно - 86). Получен лучше результат, чем в модели K-NN.Таким образом, на данном этапе отдаем предпочтение данной моделе.